### PR TITLE
Unverified Email: Interface

### LINKED ISSUE #5086: Account Email: Handle Unverified Scenario
#### Details:
Upon Account Email change, we need to display the `Pending Confirmation` legend, and allow the user to cancel the pending OP.

Ref.: #4823

screen shot 2016-01-07 at 11 11 12 am

### PR DESCRIPTION
**Pull request by jleandroperez**
Closes #5086

#### Details:

- **ErrorAnimator** shall be known, from now on, as **NoticeAnimator**.
- **ImmuTableViewController.errorMessage** was renamed to **noticeMessage**.
- **NoticeAnimator** will now dynamically resize whenever the text height changes.
- **SettingsTextViewController** was extended to support *Action Buttons*.
- **AccountSettingsService.settings** got a RxSwift patch. << That may be the most important change!

#### Scenario A: Happy Revert Path
1. Launch WPiOS and open `Me` > `Account Settings`
2. Change your email
3. `Account Settings` should display a blue notice, on top, indicating that there's a pending email change.
4. Tap over `Email`
5. The user should see a `Revert` button at the bottom
6. Tap over `Revert`

As a result, the notice should be gone. Please, note that step (3) may require a "Back & Forward" due to the known Rx race condition.

#### Scenario B: Happy Confirmed Path
Exact same steps as above. But replace step (5) as follows:

5. Open your email and click over the confirm link.

Relaunch the app: the notice should be gone.

#### Scenario C: Connectivity Loss
1. Assuming you've got a pending email change, launch the app and open Account Settings
2. Disable networking

As a result, the Notice should resize to accomodate for the different text height. Please, test rotations + iPad as well.

Needs review: @koke 
Thanks *a whole lot* for all of your patience and support with Rx!!!!

### CODE REVIEW

File: WordPress/Classes/Services/AccountSettingsService.swift [hunk 1]
```diff
@@ -148,7 +148,7 @@ class AccountSettingsService {
         // This was the simplest implementation. If performance is an issue, we could try
         // adding `distinctUntilChanged` or `filter` on the notification userInfo and only
         // emit if the changed objects include the observed account.
-        return notificationObserver.map(getSettings).startWith(getSettings())
+        return Observable.just(getSettings()).concat(notificationObserver.map(getSettings))
```
Comment (koke) [inline WordPress/Classes/Services/AccountSettingsService.swift]: This still has the same bug, as it's just a different way to write the same thing. What we want for that initial value is to call `getSettings` when the observable is subscribed to, not when it's created. I think the simplest way is to use `deferred`. 

I'm pushing ff9c412 with a fix, and hopefully adding a bit more clarity to the whole method.

File: WordPress/Classes/Models/AccountSettings.swift [hunk 2]
```diff
@@ -32,4 +32,12 @@ extension AccountSettings {
...
         language = managed.language
     }
+    
+    var emailForDisplay : String {
+        if let unwrappedPendingEmail = emailPendingAddress where unwrappedPendingEmail.isEmpty == false {
+            return unwrappedPendingEmail
+        }
+
+        return email
+    }
```
Comment (koke) [inline WordPress/Classes/Models/AccountSettings.swift]: I think this would do the same thing and I find it a bit clearer:

```swift
return emailPendingAddress?.nonEmptyString() ?? email
```

`nonEmptyString` basically turns an empty string into a `nil`

File: WordPress/Classes/Utility/NoticeAnimator.swift [hunk 3]
```diff
@@ -0,0 +1,150 @@
+import UIKit
+import WordPressShared
+
+/// NoticeAnimator is a helper class to animate error messages.
+///
+/// The error messages show at the top of the target view, and are meant to
```
Comment (koke) [inline WordPress/Classes/Utility/NoticeAnimator.swift]: "error messages" -> "notices"

File: WordPress/Classes/ViewRelated/Tools/SettingsTextViewController.m [hunk 4]
```diff
@@ -228,13 +281,16 @@ - (void)updateModeSettings:(SettingsTextModes)newMode
 {
     BOOL requiresSecureTextEntry = NO;
     UIKeyboardType keyboardType = UIKeyboardTypeDefault;
+    UITextAutocapitalizationType autocapitalizationType = UITextAutocapitalizationTypeWords;
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Tools/SettingsTextViewController.m]: Why was this added? I don't think that's the desired default behaviour

**Comment by koke** [comment]
`SettingsTextViewController` used to be an easy to understand view controller. It had a text field and an explanation, and when dismissed it sent the text value back. A sort of [prompt](http://www.w3schools.com/jsref/met_win_prompt.asp). This is something else.

Now it's that, but also:
- supports different text modes
- validates the input based on those modes
- optionally shows (animates?) a notice on the top
- optionally shows an "action" button
- optionally adds Done/Cancel to the navigation bar

When I look at its header in isolation, I'm not sure what notice and action are.

I think part of the problem is that UIKit isn't great at splitting the UI into smaller components (without a lot of boilerplate), but there's still room for improvement within those limits.

There are already a bunch of other prompt-style controllers (like `SettingsMultiTextViewController`). What happens when they also need to show a notice? Or when they also need to be presented modally and add Done/Cancel.

This looks like we might need a Settings Container that handles all of that common stuff, so the other VCs can go back to just dealing with asking the user for input.

If VC containment was a bit more straightforward to use, I'd love to see us having more specific classes rather than adding configuration (e.g. `ValidatedEmailPromptWithActionViewController`, or `PasswordPromptViewController`).

File: WordPress/Classes/Services/AccountSettingsService.swift (hunk 1 above)
Comment (jleandroperez) [inline WordPress/Classes/Services/AccountSettingsService.swift]: Thank you!!

File: WordPress/Classes/Models/AccountSettings.swift (hunk 2 above)
Comment (jleandroperez) [inline WordPress/Classes/Models/AccountSettings.swift]: Nice touch!

File: WordPress/Classes/ViewRelated/Tools/SettingsTextViewController.m (hunk 4 above)
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Tools/SettingsTextViewController.m]: You're right, the default is actually `UITextAutocapitalizationTypeSentences`, i'll switch to that.

This was added because... email addresses were being capitalized. Just handling the default behavior as well.

**Comment by jleandroperez** [comment]
@koke i do have second thoughts on `SettingsTextViewController` (in fact... i've given it some thought). The class is a bit overloaded.

Problem is, as you've mentioned, containment: 
- Adding this new functionality via subclassing involves, literally, hacking the `UITableViewDataSource` methods. Which is even worse than having that code within `SettingsTextViewController`.
- Simply creating a new ViewController for this task implies duplicating a huge portion of `SettingsTextViewController`'s current code.

Perhaps the right answer to this is: *components*. 
- A `SetupComponent` (or a better name) that could receive the TableView, and maybe, add the notice
- Still, adding the ActionButton definitely requires tampering with the TableViewDataSource protocol. Which (may be) solved via another component that gets to "add" extra rows, but sounds even more complicated than the current implementation.

Definitely open to suggestions!. All of the comments addressed, except for the `SettingsTextViewController` thoughts.

As always, thanks for the review sir!!

**Comment by jleandroperez** [comment]
@koke follow up to the architecture discussion: `PromptViewController` implemented in #5147.

**Comment by koke** [comment]
OK, let's do this. I'm not too happy about a couple things:

- Notices: we should start using the notice pattern in a lot more places to notify the user. We need a much better way to integrate them, so let's think about that.
- "Action": one thing that bugs me is that there are three separate action-related properties, that can be in an invalid state (e.g. displaysActionButton=YES, actionText=nil). But mostly it's that it doesn't really fit the previously simple description of SettingsTextViewController.

In the end, I think we need to start thinking about ways to componentize our UI better.

For now,
:shipit:

**Comment by jleandroperez** [comment]
Thank you @koke !!!
